universe/Condition.cpp

```diff
@@ -9956,6 +9965,238 @@ unsigned int Not::GetCheckSum() const {
+    // or if no candidates match A but any candidate matches B, then all candidates that do not match B are matched,
+    // or if no candidate matches A or B but any candidate matches C, then all candidates that do not match C are matched.
+    // If no candidate matches A, B, or C, then all candidates are matched.
+    if (search_domain == NON_MATCHES) {
+        // Check each operand condition on objects in the input matches and non_matches sets, until an operand condition matches an object.
+        // If an operand condition is selected, apply it to the input non_matches set, moving matching candidates to matches.
+        // If no operand condition is selected, because no candidate is matched by any operand condition, then do nothing.
+        ObjectSet temp_objects;
+        temp_objects.reserve(std::max(matches.size(),non_matches.size()));
+
+        for (auto& operand : m_operands) {
+            operand->Eval(local_context, temp_objects, non_matches, NON_MATCHES);
+            if (!temp_objects.empty()) {
+                // Select the current operand condition. Apply it to the NON_MATCHES candidate set.
+                // We alread did the application, so we use the results
+                FCMoveContent(temp_objects, matches);
+                return;
+            }
+            // Check if the operand condition matches an object in the other input set
+            operand->Eval(local_context, matches, temp_objects, MATCHES);
+            if (!matches.empty()) {
+                // Select the current operand condition. Apply it to the NON_MATCHES candidate set.
+                // We already did the application before, but there were no matches.
+                // restore state before checking the operand
+                FCMoveContent(temp_objects, matches);
+                return;
+            }
+
+            // restore state before checking the operand
+            FCMoveContent(temp_objects, matches);
+            // try the next operand
+        }
+
+        // No operand condition was selected, nothing changed, done.
+    } else /*(search_domain == MATCHES)*/ {
+        // Check each operand condition on objects in the input matches and non_matches sets, until an operand condition matches an object.
+        // If an operand condition is selected, apply it to the input matches set, moving non-matching candidates to non_matches.
+        // If no operand condition is selected, because no candidate is matched by any operand condition, then move all of the input matches into non_matches.
+        ObjectSet temp_non_matches;
+        temp_non_matches.reserve(non_matches.size());
+        matches.reserve(std::max(matches.size(),non_matches.size()));
+
+        for (auto& operand : m_operands) {
+            // Apply the current operand optimistically. Select it if there are any matching objects in the input sets 
+            operand->Eval(local_context, matches, temp_non_matches, MATCHES);
```

Comment:
If you evaluated this with the `NON_MATCHES` domain and `matches` as the second / source input set and `temp_objects` as the first / destination, then matches would be left containing the stuff that doesn't match the operand condition... You would still need to swap the contents if nothing in `non_matches` was matched by that operand condition, but it could save a temp set, I think?

Reply:
I think you are right. Looks i even got the implementation right first try and it looks more structured. There is one extra move though, but it is the happy path anyway (so just a single time happening) and it is based on moving, so i think the cleaner implementation is worth it.